src/webdav_fs.rs

```diff
@@ -10,83 +10,165 @@ use dav_server::{
-        }
+        Box::pin(async move {
+            if !self.is_path_allowed(path).await {
+                Err(DavFsError::NotFound)
+            } else {
+                self.local.open(path, options).await
+            }
+        })
     }
 
     fn read_dir<'a>(
         &'a self,
         path: &'a DavPath,
         meta: DavReadDirMeta,
     ) -> DavFsFuture<'a, DavFsStream<Box<dyn DavDirEntry>>> {
-        if self.show_hidden {
-            self.local.read_dir(path, meta)
-        } else if !path_has_hidden_components(path) {
-            Box::pin(self.local.read_dir(path, meta).map_ok(|stream| {
-                let dyn_stream: DavFsStream<Box<dyn DavDirEntry>> =
-                    Box::pin(stream.filter(|entry| {
-                        ready(match entry {
-                            Ok(e) => !e.name().starts_with(b"."),
-                            _ => false,
-                        })
-                    }));
-                dyn_stream
-            }))
-        } else {
-            Box::pin(ready(Err(DavFsError::NotFound)))
-        }
+        Box::pin(async move {
+            if !self.is_path_allowed(path).await {
+                return Err(DavFsError::NotFound);
+            }
+
+            if self.show_hidden && !self.no_symlinks {
+                return self.local.read_dir(path, meta).await;
+            }
+
+            let dav_path = path.as_rel_ospath();
+            let base_path = self.base_path.join(dav_path);
+            let show_hidden = self.show_hidden;
+            let no_symlinks = self.no_symlinks;
+
+            let stream = self.local.read_dir(path, meta).await?;
+
+            let filtered = stream.filter_map(move |entry_res| {
+                let base_path = base_path.clone();
+                async move {
+                    match entry_res {
+                        Ok(e) => {
+                            if !show_hidden && e.name().starts_with(b".") {
+                                return None;
+                            }
+                            if no_symlinks {
+                                let name = e.name();
+                                #[cfg(not(target_os = "windows"))]
+                                let os_string = OsStr::from_bytes(&name);
+                                #[cfg(target_os = "windows")]
+                                let os_string: &OsStr =
+                                    std::str::from_utf8(&name).unwrap().as_ref();
+                                let entry_path = base_path.join(os_string);
+                                if let Ok(md) = fs::symlink_metadata(&entry_path).await {
+                                    if md.file_type().is_symlink() {
+                                        return None;
+                                    }
+                                }
+                            }
+                            Some(Ok(e))
+                        }
+                        Err(e) => Some(Err(e)),
```

Comment:
Hmmm is this right? Won't this result in weird entries? Also, should we log here?